group-coordinator/src/test/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilderTest.java

```diff
@@ -0,0 +1,679 @@
+                subscribedTopics,
+                targetAssignment.partitions()
+            );
+        }
+
+        public TargetAssignmentBuilder.TargetAssignmentResult build() {
+            // Prepare expected member specs.
+            Map<String, AssignmentMemberSpec> memberSpecs = new HashMap<>();
+            members.forEach((memberId, member) -> {
+                memberSpecs.put(memberId, assignmentMemberSpec(
+                    member,
+                    targetAssignments.getOrDefault(memberId, Assignment.EMPTY)
+                ));
+            });
+
+            updatedMembers.forEach((memberId, updatedMemberOrNull) -> {
+                if (updatedMemberOrNull == null) {
+                    memberSpecs.remove(memberId);
+                } else {
+                    memberSpecs.put(memberId, assignmentMemberSpec(
+                        updatedMemberOrNull,
+                        targetAssignments.getOrDefault(memberId, Assignment.EMPTY)
+                    ));
+                }
+            });
+
+            Map<Uuid, AssignmentTopicMetadata> topicMetadata = new HashMap<>();
+            subscriptionMetadata.forEach((topicName, metadata) -> {
+                topicMetadata.put(metadata.id(), new AssignmentTopicMetadata(metadata.numPartitions()));
+            });
+
+            AssignmentSpec assignmentSpec = new AssignmentSpec(
+                memberSpecs,
+                topicMetadata
+            );
+
+            // We use `any` here to always return an assignment but use `verify` later on
+            // to ensure that the input was correct.
+            when(assignor.assign(any())).thenReturn(new GroupAssignment(assignments));
+
+            // Create and populate the assignment builder.
+            TargetAssignmentBuilder builder = new TargetAssignmentBuilder(groupId, groupEpoch, assignor)
+                .withMembers(members)
+                .withSubscriptionMetadata(subscriptionMetadata)
+                .withTargetAssignments(targetAssignments);
+
+            updatedMembers.forEach((memberId, updatedMemberOrNull) -> {
+                if (updatedMemberOrNull != null) {
+                    builder.withUpdatedMember(memberId, updatedMemberOrNull);
+                } else {
+                    builder.withRemoveMembers(memberId);
+                }
+            });
+
+            TargetAssignmentBuilder.TargetAssignmentResult result = builder.build();
+
+            verify(assignor, times(1)).assign(assignmentSpec);
+
+            return result;
+        }
+    }
+
+    @Test
+    public void testEmpty() {
+        TargetAssignmentBuilderTestContext context = new TargetAssignmentBuilderTestContext(
+            "my-group",
+            20
+        );
+
+        TargetAssignmentBuilder.TargetAssignmentResult result = context.build();
+        assertEquals(newTargetAssignmentEpochRecord(
+            "my-group",
+            20
+        ), result.records().get(result.records().size() - 1));
```

Comment:
is the size non-deterministic? isn't the size 1 since we don't expect any TargetAssignmentMember records?

Reply:
Correct. In this case, it is even better to use `Collections.singletonList` and to verify the records list.